### PR TITLE
SI-9387 Fix VerifyError introduced by indylambda

### PR DESCRIPTION
**Pull request by retronym**
As with regular `Apply`-s, we should compute the generated type
based on the function's type, rather than the expected type.

In the test case, the expected type was void. Now, we correctly
use the generated type of `scala/Function1`, which is enough
to generate a subsequent POP instruction.

The tree shape involved was:

```
arg0 = {
  {
    $anonfun()
  };
  scala.runtime.BoxedUnit.UNIT
}
```

Review by @lrytz

Targetting to 2.12.x to expediate M2, but we should backport
this (along with one two other recent changes) to 2.11.x.

### CODE REVIEW

**Comment by retronym** [comment]
Assigned to the 2.12.0-M2, as this is a blocker for the release, /cc @SethTisue

**Comment by lrytz** [comment]
here are two smaller test cases to reproduce the error, maybe we could add them?

```
object T {
  val f: Unit = () => ()
  println(f)
}

object U {
  def f[T](t: T): T = t
  f[Unit](() => ())
}
```

**Comment by lrytz** [comment]
otherwise LGTM

**Comment by retronym** [comment]
Added.

**Comment by lrytz** [comment]
i don't see them.. :)

**Comment by retronym** [comment]
Third time lucky.

**Comment by lrytz** [comment]
LGTM

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala [hunk 1]
```diff
@@ -632,10 +632,11 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
             case _ =>
               abort(s"Cannot instantiate $tpt of kind: $generatedType")
           }
-        case Apply(_, args) if app.hasAttachment[delambdafy.LambdaMetaFactoryCapable] =>
+        case Apply(fun, args) if app.hasAttachment[delambdafy.LambdaMetaFactoryCapable] =>
```
Comment (viktorklang) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:635]: Everything becomes better after fun is applied.

**Pull request merged**